### ISSUE TITLE
Mismatch between test set metrics during training and evaluation

### ISSUE THREAD

**Issue by dimokaramanlis** [DeepLabCut/DeepLabCut#3083]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating System

Windows 10

### DeepLabCut version

3.0.0rc10

### What engine are you using?

pytorch

### DeepLabCut mode

multi animal

### Device type

NVIDIA RTX A4000

### Bug description 🐛

To DLC developers,

I have been using DLC to track two mice (12 keypoints each) very successfully in the past using a resnet50 core with tensorflow.

Since moving to pytorch, I trained new models using the same datasets so that we can have longer-term support. In the latest version (rc10), I keep coming across a (large) mismatch between the metrics reported during model training and model evaluation. 

For example (see file below), my model (dlcrnet) is being trained normally and both train and test set errors are decreasing. The model of epoch 150 is selected as "best" and has a test rmse of ~4.4. See below:

[learning_stats.csv](https://github.com/user-attachments/files/22075605/learning_stats.csv)

When I'm running the evaluation later, the same model only gets a test rmse of 10, which is much worse than what is reported during training. Visual inspection of the predictions is also off. See below.

[DLC_DlcrnetStride16Ms5_TwoMicePredictionAug29shuffle1_snapshot_best-150-results.csv](https://github.com/user-attachments/files/22075603/DLC_DlcrnetStride16Ms5_TwoMicePredictionAug29shuffle1_snapshot_best-150-results.csv)

Do you have an idea of what may be the issue? I have never had a similar issue with tensorflow models, or pytorch models trained with the rc9 version. As as side note, model iinference with rc9 was much slower.

Thank you very much for the amazing software.

Best,
Dimos

### Relevant log output

```shell

```

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by maximpavliv** [comment]
Hi @dimokaramanlis ,

Thanks for reporting this.

Could you please share your _pytorch_config.yaml_ and your _train.txt_ files as well ? Thanks in advance!

**Comment by dimokaramanlis** [comment]
Yes of course. Here is a list of all relevant files from a new run with resnet50. I still have the same issue. Snapshot at 200 metrics do not match the evaluation metrics.

[CombinedEvaluation-results.csv](https://github.com/user-attachments/files/22092409/CombinedEvaluation-results.csv)
[learning_stats.csv](https://github.com/user-attachments/files/22092416/learning_stats.csv)
[pytorch_config.yaml](https://github.com/user-attachments/files/22092418/pytorch_config.yaml)
[train.txt](https://github.com/user-attachments/files/22092417/train.txt)

**Comment by maximpavliv** [comment]
Hey @dimokaramanlis ,

Sorry for getting back so late to you.

I think I found how to reproduce the issue! To confirm my hypothesis, could you try running evaluation again, but specify `device="cpu"` when calling `deeplabcut.evaluate_network`? Do you then get the same metrics as the ones you saw in the training logs?

**Comment by dimokaramanlis** [comment]
Your hypothesis is correct! The metrics now match:

Evaluation results file: DLC_Resnet50_DyadTorchSep3shuffle1_snapshot_best-170-results.csv
Evaluation results for DLC_Resnet50_DyadTorchSep3shuffle1_snapshot_best-170-results.csv (pcutoff: 0.6):
train rmse             2.10
train rmse_pcutoff     2.08
train mAP             99.20
train mAR             99.60
test rmse              4.16
test rmse_pcutoff      4.14
test mAP              98.90
test mAR              99.05
Name: (0.95, 1, 170, -1, 0.6), dtype: float64

Then I guess I trained the model on the cpu and somehow the weights don't properly transfer to the gpu (which I typically use for inference)?

**Comment by maximpavliv** [comment]
Hi @dimokaramanlis ,

Thank you for confirming the hypothesis! I’m quite certain that your model was trained on the GPU — this issue actually comes from our side. Starting in DeepLabCut version **3.0.0rc9**, we introduced a change aimed at speeding up GPU inference, but thanks to your report we’ve just realized that this change negatively affects the accuracy during evaluation/inference on the GPU.

The good news is that you **don’t need to retrain your model** — the problem only impacts inference, not training. We’ll work on a fix next week. In the meantime, if you’d like inference to match the training accuracy, you can either
- run evaluation with `device="cpu"` (running video analyis may be slow), or
- temporarily downgrade to **v3.0.0rc8**, which will give the same accuracy with the GPU as in training.

Thanks again for catching this and helping us track it down!

Best,
Maxim

**Comment by dimokaramanlis** [comment]
Thanks for the suggestions, for now I went back to using my previous tensorflow model, and I'll be waiting for the fix!

Best,
Dimos

**Comment by maximpavliv** [comment]
Hi @dimokaramanlis,

I’m happy to let you know that we’ve merged the Pull Request with the fix for this issue ([#3105](https://github.com/DeepLabCut/DeepLabCut/pull/3105)) and just [released](https://pypi.org/project/deeplabcut/3.0.0rc13/) a new version of DeepLabCut that resolves the problem you reported. 🎉

This release restores the correct inference accuracy, so you don’t need to retrain your model or make any changes — your evaluation and video analysis performance will be back to normal after updating.

Thank you again for your highly valuable feedback — reports like yours are incredibly useful to help us improve DeepLabCut.

I’ll close this issue for now, but please feel free to re-open it if you encounter any further problems.

Best,
Maxim
